### DISCUSSION TITLE
How to keep the Indicators and Markers when restoring solutions

### DISCUSSION THREAD

**Discussion by stonebeta** [idaholab/moose#18962] [Q&A General] [answered]
Hello 
In the file Transient.C, line 409-415, when solve is not converged:
```C
  else
  {
    _problem.restoreMultiApps(EXEC_TIMESTEP_BEGIN, true);
    _problem.restoreMultiApps(EXEC_TIMESTEP_END, true);
    _time_stepper->rejectStep();
    _time = _time_old;
  }
```
I add  _problem.computeIndicators(), _problem.computeMarkers() and _problem.adaptMesh() here, like this:
```C
 else
  {
    if (mesh_error())
    {
      _problem.computeIndicators();
      _problem.computeMarkers();
    }

    _problem.restoreMultiApps(EXEC_TIMESTEP_BEGIN, true);
    _problem.restoreMultiApps(EXEC_TIMESTEP_END, true);
    _time_stepper->rejectStep();
    _time = _time_old;

    if (mesh_error())
      _problem.adaptMesh();
  }
```
What I want to do is:
1. Compute the indicators and markers of the unconverged step.
2. Then restore everything except Indicators and markers.
3. Refine the mesh according to the indicators and markers above.

But indicators and markers seem to be restored as well.
Is there any way to keep them？

**Comment by GiudGiud** [comment, marked as answer]
I don't think there is an option to restore only part of the system.

You could save a copy of that part of solution vector then set it again.

```
    // Save variable previous values
    NumericVector<Number> old = _nl.getVector(tag_id   OR   variable name);
```

Are you trying to refine the multiapp or the main app btw

**Reply by stonebeta** [reply]
I am trying to refine the multiapp and by level_set

**Reply by GiudGiud** [reply]
I dont understand what you mean by `and by level_set`.
Ah saving the marker and indicator from the multiapp here may be tough. You'd need to transfer them or something. Even then i dont know that it happens in the right order.

**Reply by stonebeta** [reply]
I use the level_set module. It seems to be able to transfer the markers. 
```
[Transfers]
  [to_markers]
    type = LevelSetMeshRefinementTransfer
    multi_app = fracture
    variable = marker
    source_variable = marker
    check_multiapp_execute_on = false
  []
.....
[Problem]
  type = LevelSetProblem
[]
```

**Reply by GiudGiud** [reply]
Ok but the markers need to be transferred before the restore, after a fail, if we are going to save the variable there.

**Reply by stonebeta** [reply]
After saving the old solution vector , how to set the new one ?

**Reply by GiudGiud** [reply]
```
NumericVector<Number> & new = _nl.getVector(tag_id   OR   variable name);
new = old;
```

**Reply by stonebeta** [reply]
There is an error
```
/home/user01/projects/moose/framework/src/executioners/Transient.C:471:54: error: cannot allocate an object of abstract type 'libMesh::NumericVector<double>'
  471 |       NumericVector<Number> old_d = _nl.getVector("d");
```

**Reply by GiudGiud** [reply]
Ah that's annoying. I think you'll need to use one of the constructors for one of the derived classes 

See the documentation for that type, see if you can find something that works
https://libmesh.github.io/doxygen/classlibMesh_1_1NumericVector.html

**Reply by stonebeta** [reply]
If I refine the mesh before restoring solutions , it is no need to save solution vectors:
``` c++
else
  {
    if (mesh_error())
    {
      _problem.computeIndicators();
      _problem.computeMarkers();
      _problem.adaptMesh();
    }

    _problem.restoreMultiApps(EXEC_TIMESTEP_BEGIN, true);
    _problem.restoreMultiApps(EXEC_TIMESTEP_END, true);
    _time_stepper->rejectStep();
    _time = _time_old;     
  }
```
But the number of the elements is changed after _problem.adaptMesh(). Is there any way to restore the solutions to the new mesh.

**Reply by stonebeta** [reply]
This seems to work:
```c++
  else
  {
    if (mesh_error())
    {
      _problem.computeIndicators();
      _problem.computeMarkers();

      _problem.adaptMesh();

      // _problem.restoreMultiApps(EXEC_TIMESTEP_BEGIN, true);
      // _problem.restoreMultiApps(EXEC_TIMESTEP_END, true);
      _time_stepper->rejectStep();
      _time = _time_old;
    }
    else
    {
      _problem.restoreMultiApps(EXEC_TIMESTEP_BEGIN, true);
      _problem.restoreMultiApps(EXEC_TIMESTEP_END, true);
      _time_stepper->rejectStep();
      _time = _time_old;
    }
  }
```
I remove the `_problem.restoreMultiApps(EXEC_TIMESTEP_BEGIN, true);` and  ` _problem.restoreMultiApps(EXEC_TIMESTEP_END, true);`. What's the use of these two lines ？ Is that safe to do this ?

**Reply by GiudGiud** [reply]
if you dont restore the MultiApp, it will not restart from where it was at the beginning, but from where it ended up before failing. 

restoring will not adapt to the new mesh imo. Refining before restoring will run into this issue.

**Reply by stonebeta** [reply]
What is the difference of `restoreMultiApps(EXEC_TIMESTEP_END, true)` and `restoreMultiApps(EXEC_TIMESTEP_BEGIN, true)` ? The `_problem.restoreMultiApps(EXEC_TIMESTEP_END, true)` here causes errors while `_problem.restoreMultiApps(EXEC_TIMESTEP_BEGIN, true)`  is ok.

**Reply by GiudGiud** [reply]
Did you mean to copy paste another name for the routine? I see `restoreMultiApps(EXEC_TIMESTEP_END, true)` twice and `_problem.restoreMultiApps(EXEC_TIMESTEP_END, true)` twice as well

**Reply by stonebeta** [reply]
Oh, I have changed that.

**Reply by GiudGiud** [reply]
So the difference is which multiapps are restored. The first one only restores the multiapps executing on timestep_end, then second one the multiapps executing at the beginning of timesteps

**Reply by GiudGiud** [reply]
any progress on this?

**Reply by GiudGiud** [reply]
Created an issue #19341 and closing this
